File: WEEK5/bj3_1197.py
import sys
import heapq

input = sys.stdin.readline


def find(x):
    if parent[x] == x:
        return x
    parent[x] = find(parent[x])
    return parent[x]


def union(x, y):
    rootX = find(x)
    rootY = find(y)
    if rootX != rootY:
        parent[rootY] = rootX


V, E = map(int, input().split())
pq = []

for _ in range(E):
    A, B, C = map(int, input().split())
    heapq.heappush(pq, (C, A, B))

parent = [i for i in range(V + 1)]
totalWeight = 0
edgeCount = 0

while pq:
    weight, start, end = heapq.heappop(pq)

    if find(start) != find(end):
        union(start, end)
        totalWeight += weight
        edgeCount += 1

    if edgeCount == V - 1:
        break

print(totalWeight)

File: WEEK5/bj3_1197.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');
let lineIndex = 0;

const [V, E] = input[lineIndex++].split(' ').map(Number);
const edges = [];

for (let i = 0; i < E; i++) {
    const [A, B, C] = input[lineIndex++].split(' ').map(Number);
    edges.push({ start: A, end: B, weight: C });
}

edges.sort((a, b) => a.weight - b.weight);

const parent = Array.from({ length: V + 1 }, (_, i) => i);

function find(x) {
    if (parent[x] === x) {
        return x;
    }
    return parent[x] = find(parent[x]);
}

function union(x, y) {
    const rootX = find(x);
    const rootY = find(y);
    if (rootX !== rootY) {
        parent[rootY] = rootX;
    }
}

let totalWeight = 0;
let edgeCount = 0;

for (const edge of edges) {
    if (find(edge.start) !== find(edge.end)) {
        union(edge.start, edge.end);
        totalWeight += edge.weight;
        edgeCount++;
    }
    if (edgeCount === V - 1) {
        break;
    }
}

console.log(totalWeight);